ENH: Add GenericSerialDevice Remote Control module

adamaji opened the pull request:
As a followup from https://github.com/PlusToolkit/PlusLib/pull/973#issuecomment-1206596033 (way later than intended), this PR adds a module similar to UltrasoundRemoteControl, but for [GenericSerialDevices](http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceGenericSerial.html). Supported functionality specifically is
- Sending data over the TX line
- Getting the CTS line
- Setting and getting the RTS line

this requires a PlusServer built with https://github.com/PlusToolkit/PlusLib/commit/a41b411bd3dfd7aadf7b6bfb17280e28e063c061.

Here's a video showing the module in action. In this setup we have the RTS line hooked up to some LEDs.

![generic-serial-control](https://user-images.githubusercontent.com/3487395/187275716-36918e49-a2da-4089-8e2a-cf37071cda4e.gif)

cc @jamesobutler for review

lassoan commented on GenericSerialDeviceRemoteControl/GenericSerialDeviceRemoteControl.py:
```diff
@@ -0,0 +1,350 @@
+class GenericSerialDeviceRemoteControl(ScriptedLoadableModule):
+    def __init__(self, parent):
+        ScriptedLoadableModule.__init__(self, parent)
+        self.parent.title = "Generic Serial Device Remote Control"
```
GenericSerialDevice is a good class name, but for a name that users see, I would skip the `Generic`. The module name would then be simply `Serial Device Remote Control`.

jamesobutler replied:
Yeah I was just about to an include a comment about the same thing.
![image](https://user-images.githubusercontent.com/15837524/187506619-f32acd14-62e1-4dc0-947d-62bb2ee78111.png)

^ A module icon for it would also be nice

adamaji replied:
Any thoughts on what the module icon would be? Is just copying the OpenIGTLink Remote icon fine?

jamesobutler replied:
I don’t think duplicating the icon of another module would add much value beyond the default extension icon.

how about something based on the serial nature such as 0/1s along a single line between two endpoints? Making the point it is serial and not parallel streams.

lassoan replied:
A serial port icon could work, too - like this:

https://iconscout.com/icon/serial-port-2653340

lassoan commented on GenericSerialDeviceRemoteControl/GenericSerialDeviceRemoteControl.py:
```diff
@@ -0,0 +1,350 @@
+        self.parent.dependencies = ["OpenIGTLinkRemote"]
+        self.parent.contributors = ["Adam Aji (SonoVol, PerkinElmer)"]
+        self.parent.helpText = """
+            Send commands to and read output from serial devices over their CTS/RTS/TX lines.
```
Mention Plus toolkit and provide a link to a documentation page with some more details.

lassoan commented on GenericSerialDeviceRemoteControl/README.md:
```diff
@@ -0,0 +1,4 @@
+# Generic Serial Device Remote Control
+Remote control interface for [GenericSerialDevices](http://perk-software.cs.queensu.ca/plus/doc/nightly/user/DeviceGenericSerial.html).
+
+Set and get values from the TX, CTS, and RTS lines on a serial device connected through OpenIGTLink using IGTL commands. Based off of UltrasoundRemoteControl.
```
Add link to an example Plus configuration file.

lassoan commented on GenericSerialDeviceRemoteControl/GenericSerialDeviceRemoteControl.py:
```diff
@@ -0,0 +1,350 @@
+        Input: callback function to run when command returns (function)
+        Output: success (bool)
+        """
+        if self._connectorNode is None or self._connectorNode.GetState() != slicer.vtkMRMLIGTLConnectorNode.StateConnected:
```
These couple of lines are repeated 4 times, consider factoring out as much as possible into a function that you call from all 4 places.

lassoan commented on GenericSerialDeviceRemoteControl/GenericSerialDeviceRemoteControl.py:
```diff
@@ -0,0 +1,350 @@
+            return False
+        logging.info("Sending GetCTS")
+        command_xml = f"""<Command Name="SerialCommand" DeviceId="{self._deviceId}" CommandName="GetCTS" />"""
+        self._command = slicer.vtkSlicerOpenIGTLinkCommand()
```
Either cleanup `self._command` (remove observer, handle the timeout/no response) or store all the in-progress commands in an array.

lassoan commented on GenericSerialDeviceRemoteControl/Testing/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+add_subdirectory(Python)
```
it's a text file, so it is good practice to terminate it with a newline

lassoan replied:
Remove the empty GenericSerialDeviceRemoteControl/Testing/Python/CMakeLists.txt file

lassoan reviewed (commented):
It looks good to me, just added some small comments.

jamesobutler reviewed (commented):
- [x] Similar to UltrasoundRemoteControl, minimize the unused Plus Parameters space by use of `addStretch` or directly with vertical spacer. This will reduce the space between the Plus Parameters and the Send Generic Command sections.
https://github.com/openigtlink/SlicerOpenIGTLink/blob/3bbf9d109e713f6da73ba742dc9b6967d74a4b65/UltrasoundRemoteControl/UltrasoundRemoteControl.py#L120

jamesobutler commented on GenericSerialDeviceRemoteControl/Resources/UI/GenericSerialDeviceRemoteControl.ui:
```diff
@@ -0,0 +1,126 @@
+    <item>
+     <widget class="ctkCollapsibleButton" name="sendCommandCollapsibleButton">
+      <property name="text">
+       <string>Send Generic Command</string>
```
How about just "Send Command"? Technically the command itself can be for something very specific.

adamaji commented:
I've added some squashable commits which I believe address all current comments. A few things:

- I deleted the entire "BUILD_TESTING" block in the top level CMakeLists.txt along with the testing subdirectory to remove the empty CMakeLists without adding an empty directory
- I went with the serial port icon for the background of the module icon
- I had some trouble with the cleanup here. The way I have it now feels kind of hacky in that I'm keeping a stage of commands which have been completed, which is then used to clean up any observers associated to those commands whenever a new command is sent. I remove the observers specifically with the (`object`, `event`, `method`) set required for `VTKObservationMixin.removeObserver`, since there isn't currently an object or event specific removal. I would like it to be as simple as UltrasoundRemoteControl where the observers are removed directly for a single method, but it seems like the assumption in that module is that only one command is running at a time (with `setInteractionInProgress`) while I don't want to keep that assumption here. The usage of the stage is so that the `CommandCompletedEvent` can also be hooked into without immediately cleaning up the command. I had also tried sending the commands back from the logic class to the widget class so the observers could be handled there, but adding observers on that end crashed Slicer whenever that code was accessed? @lassoan would you have any additional suggestions here?

jamesobutler reviewed (commented):
For reference an image with the new icon for Serial Device Remote Control (remote control + Serial port) which is similar to the Ultrasound Remote Control module (remote control + Ultrasound probe)
![image](https://user-images.githubusercontent.com/15837524/188695235-0619dab9-717e-4807-bce6-f70adda73901.png)

Latest version of the module widget:
![image](https://user-images.githubusercontent.com/15837524/188695591-9e5e4249-f3b6-4980-bc76-401169b28130.png)

adamaji commented:
I've squashed down all the fixup commits. @Sunderlandkyl would you be able to review / merge here?

Sunderlandkyl commented:
Looks, good to me!

adamaji commented:
Great thanks!
